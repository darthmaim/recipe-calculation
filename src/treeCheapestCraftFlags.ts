import { RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'

// Set the craft flags based on the cheapest price
export function treeCheapestCraftFlags(
  tree: RecipeTreeWithPrices,
  forceBuyItems: Array<number>
): RecipeTreeWithCraftFlags {
  // Craft the item if it can't be bought or it is cheaper to craft
  const shouldForceBuy = forceBuyItems.indexOf(tree.id) !== -1
  const isUsed = tree.usedQuantity !== 0
  const craftIsCheaper =
    typeof tree.craftPrice !== 'undefined' && (!tree.buyPrice || tree.craftPrice < tree.buyPrice)

  const craft = !!tree.components && !shouldForceBuy && isUsed && craftIsCheaper

  if (!tree.components) {
    return { ...tree, components: undefined, craft }
  }

  // Adjust the flags for all tree's sub-recipes
  const components = tree.components.map((component) =>
    treeCheapestCraftFlags(component, forceBuyItems)
  )

  return { ...tree, components, craft }
}
