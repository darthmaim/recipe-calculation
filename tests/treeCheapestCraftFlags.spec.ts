import treeCheapestCraftFlags from '../src/treeCheapestCraftFlags'

describe('treeCheapestCraftFlags', () => {
  it('sets the cheapest craft flags', () => {
    let recipeTree = {
      id: 1,
      totalQuantity: 1,
      craftPrice: 123,
      components: [
        { id: 2, totalQuantity: 1, craftPrice: 123, buyPrice: 1 },
        { id: 3, totalQuantity: 2, craftPrice: 123 },
        { id: 4, totalQuantity: 2, craftPrice: 123, buyPrice: 555 },
        { id: 5, totalQuantity: 2, craftPrice: 1, buyPrice: 555 },
      ],
    }

    let calculatedTree = treeCheapestCraftFlags(recipeTree, [5])
    expect(calculatedTree).toEqual({
      craft: true,
      craftPrice: 123,
      id: 1,
      totalQuantity: 1,
      components: [
        {
          buyPrice: 1,
          craft: false,
          craftPrice: 123,
          id: 2,
          totalQuantity: 1,
        },
        {
          craft: true,
          craftPrice: 123,
          id: 3,
          totalQuantity: 2,
        },
        {
          buyPrice: 555,
          craft: true,
          craftPrice: 123,
          id: 4,
          totalQuantity: 2,
        },
        {
          buyPrice: 555,
          craft: false,
          craftPrice: 1,
          id: 5,
          totalQuantity: 2,
        },
      ],
    })
  })
})
