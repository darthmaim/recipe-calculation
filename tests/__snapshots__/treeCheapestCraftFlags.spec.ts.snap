// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeCheapestCraftFlags sets the cheapest craft flags 1`] = `
Object {
  "buyPrice": false,
  "buyPriceEach": false,
  "components": Array [
    Object {
      "buyPrice": 1,
      "buyPriceEach": 1,
      "components": Array [],
      "craft": false,
      "craftPrice": 123,
      "decisionPrice": 1,
      "disciplines": Array [],
      "id": 2,
      "min_rating": null,
      "output": 1,
      "quantity": 1,
      "totalQuantity": 1,
      "type": "Recipe",
      "usedQuantity": 1,
    },
    Object {
      "buyPrice": false,
      "buyPriceEach": false,
      "components": Array [],
      "craft": true,
      "craftPrice": 123,
      "decisionPrice": 123,
      "disciplines": Array [],
      "id": 3,
      "min_rating": null,
      "output": 1,
      "quantity": 2,
      "totalQuantity": 2,
      "type": "Recipe",
      "usedQuantity": 2,
    },
    Object {
      "buyPrice": 555,
      "buyPriceEach": 277,
      "components": Array [],
      "craft": true,
      "craftPrice": 123,
      "decisionPrice": 123,
      "disciplines": Array [],
      "id": 4,
      "min_rating": null,
      "output": 1,
      "quantity": 2,
      "totalQuantity": 2,
      "type": "Recipe",
      "usedQuantity": 2,
    },
    Object {
      "buyPrice": 555,
      "buyPriceEach": 277,
      "components": Array [],
      "craft": false,
      "craftPrice": 1,
      "decisionPrice": 1,
      "disciplines": Array [],
      "id": 5,
      "min_rating": null,
      "output": 1,
      "quantity": 2,
      "totalQuantity": 2,
      "type": "Recipe",
      "usedQuantity": 2,
    },
  ],
  "craft": true,
  "craftPrice": 123,
  "decisionPrice": 123,
  "disciplines": Array [],
  "id": 1,
  "min_rating": null,
  "output": 1,
  "quantity": 1,
  "totalQuantity": 1,
  "type": "Recipe",
  "usedQuantity": 1,
}
`;
